Switch flex scrollers to allow infinite endward overflow, plus startward overflow limited to in-flow alignment

fantasai opened the pull request:
#### ce08f3245391471d182f822c27d11345dc5e0a24
<pre>
Switch flex scrollers to allow infinite endward overflow, plus startward overflow limited to in-flow alignment
<a href="https://bugs.webkit.org/show_bug.cgi?id=275209">https://bugs.webkit.org/show_bug.cgi?id=275209</a>
<a href="https://rdar.apple.com/problem/129343181">rdar://problem/129343181</a>

Reviewed by Sammy Gill.

Switches allowedLayoutOverflow to allow infinite overflow in the end direction
(per writing mode), and to allow limited overflow based on in-flow alignment.
This matches Flexbox to the logic used in Grid and Block layout.

* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-overflow-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-overflow-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-overflow-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-overflow-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-overflow-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-overflow-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-overflow-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-overflow-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-overflow-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-overflow-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-overflow-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-overflow-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-overflow-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-overflow-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-overflow-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-overflow-002.html: Added.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::resolveLeftRightAlignment): Add convenience method.
(WebCore::initialJustifyContentOffset): Use convenience method.
(WebCore::contentAlignmentStartOverflow): Add convenience method.
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren): Stash alignment overflow.
(WebCore::RenderFlexibleBox::alignFlexLines): Stash alignment overflow.
(WebCore::RenderFlexibleBox::allowedLayoutOverflow const): Use alignment overflow for limits.
* Source/WebCore/rendering/RenderFlexibleBox.h: Add storage for alignment overflow.

Canonical link: <a href="https://commits.webkit.org/279992@main">https://commits.webkit.org/279992@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf4a52433f207dd3d0ca885f1e46ce6cfaeff97f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55430 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34751 "Built successfully") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7890 "Failed to checkout and rebase branch from PR 29679") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58418 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5864 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57557 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42370 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6052 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44637 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4012 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57457 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32661 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/7890 "Failed to checkout and rebase branch from PR 29679") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25765 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29446 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/7890 "Failed to checkout and rebase branch from PR 29679") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4008 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51382 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/7890 "Failed to checkout and rebase branch from PR 29679") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60008 "Built successfully") | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30578 "Failed to checkout and rebase branch from PR 29679") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5486 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52067 "Passed tests") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31662 "Failed to checkout and rebase branch from PR 29679") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/7890 "Failed to checkout and rebase branch from PR 29679") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51530 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12267 "Built successfully and passed tests") | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32743 "Failed to checkout and rebase branch from PR 29679") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31414 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1805,6 +1805,16 @@ bool RenderFlexibleBox::resolveFlexibleLengths(FlexSign flexSign, FlexItems& fle
 
+inline ContentPosition resolveLeftRightAlignment(ContentPosition position, const RenderStyle& style, bool isReversed)
+{
+    if (position == ContentPosition::Left || position == ContentPosition::Right) {
```
This function seems to indicate by the name and the first `if` guard here that passing in a `ContentPosition` that is not left/right is an error. Should we maybe also `ASSERT` here to indicate that and make sure that the caller only passes in valid values here?

fantasai replied:
The function can handle any value. It only changes the left/right values.

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1805,6 +1805,16 @@ bool RenderFlexibleBox::resolveFlexibleLengths(FlexSign flexSign, FlexItems& fle
     return !totalViolation;
 }
 
+inline ContentPosition resolveLeftRightAlignment(ContentPosition position, const RenderStyle& style, bool isReversed)
```
I think it would be nice if the function name stated what it was resolving the passed in alignment values to

fantasai replied:
I think `resolveLeftRightAlignmentToStartEnd` would be more confusing than leaving it blank, actually, because then it reads as if maybe we're mapping left to start and right to end.

The contract here is that left/right are mapping to something other than left/right, so I think it's best left as-is?

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2181,6 +2187,35 @@ bool RenderFlexibleBox::childHasPercentHeightDescendants(const RenderBox& render
+        return -availableFreeSpace;
+    default:
+        ASSERT((distribution == ContentDistribution::Default && position == ContentPosition::Normal) // Normal alignment.
+            || position == ContentPosition::FlexStart
```
Isn't `FlexStart` already covered in one of the cases above?

fantasai replied:
Right. It wasn't in a previous iteration, but I should fix that. :)

fantasai committed:
Switch flex scrollers to allow infinite endward overflow, plus startward overflow limited to in-flow alignment
https://bugs.webkit.org/show_bug.cgi?id=275209
rdar://problem/129343181

Reviewed by Sammy Gill.

Switches allowedLayoutOverflow to allow infinite overflow in the end direction
(per writing mode), and to allow limited overflow based on in-flow alignment.
This matches Flexbox to the logic used in Grid and Block layout.

* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-overflow-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-overflow-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-overflow-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-overflow-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-overflow-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-overflow-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-overflow-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-overflow-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-overflow-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-overflow-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-overflow-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-overflow-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-overflow-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-overflow-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-overflow-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-overflow-002.html: Added.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::resolveLeftRightAlignment): Add convenience method.
(WebCore::initialJustifyContentOffset): Use convenience method.
(WebCore::contentAlignmentStartOverflow): Add convenience method.
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren): Stash alignment overflow.
(WebCore::RenderFlexibleBox::alignFlexLines): Stash alignment overflow.
(WebCore::RenderFlexibleBox::allowedLayoutOverflow const): Use alignment overflow for limits.
* Source/WebCore/rendering/RenderFlexibleBox.h: Add storage for alignment overflow.

Canonical link: https://commits.webkit.org/279992@main
